src/object.rs

```diff
@@ -866,53 +880,112 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
-            w
+            let weak = WeakRef {
+                inner: Arc::new(WeakRefInner(mem::zeroed())),
+                thread: Some(thread::current().id()),
+                marker: PhantomData,
+            };
+            gobject_ffi::g_weak_ref_init(mut_override(&weak.inner.0), self.to_glib_none().0);
+            weak
         }
-
     }
 }
 
-pub struct WeakRef<T: IsA<Object> + ?Sized>(Box<gobject_ffi::GWeakRef>, PhantomData<*const T>);
+// This trait is only necessary because specialization with default impl is unstable.
+pub trait SyncObjectExt: ObjectExt {
+    /// Creates a new weak reference to this object.
+    ///
+    /// Returned reference can be upgraded back into strong reference from any thread.
+    ///
+    /// See also [`downgrade`][downgrade] for objects that aren't `Sync`.
+    ///
+    /// [downgrade]: trait.ObjectExt.html#tymethod.downgrade
+    fn downgrade_sync(&self) -> WeakRef<Self>;
+}
 
-impl<T: IsA<Object> + StaticType + UnsafeFrom<ObjectRef> + Wrapper + ?Sized> WeakRef<T> {
-    pub fn upgrade(&self) -> Option<T> {
+impl<T: IsA<Object> + ObjectExt + Sync> SyncObjectExt for T {
+    fn downgrade_sync(&self) -> WeakRef<T> {
         unsafe {
-            let ptr = gobject_ffi::g_weak_ref_get(mut_override(&*self.0));
-            if ptr.is_null() {
-                None
-            } else {
-                let obj: Object = from_glib_full(ptr);
-                Some(T::from(obj.into()))
-            }
+            let weak = WeakRef {
+                inner: Arc::new(WeakRefInner(mem::zeroed())),
+                thread: None,
+                marker: PhantomData,
+            };
+            gobject_ffi::g_weak_ref_init(mut_override(&weak.inner.0), self.to_glib_none().0);
+            weak
         }
     }
 }
 
-impl<T: IsA<Object> + ?Sized> Drop for WeakRef<T> {
+#[derive(Clone, Debug)]
+pub struct WeakRef<T: ?Sized> {
+    /// The only way to add another weak reference to GObject WeakRef is to upgrade it first. This
+    /// wouldn't be thread-safe for types that aren't `Sync`, so instead we add another layer of
+    /// indirection.
+    inner: Arc<WeakRefInner>,
+    thread: Option<thread::ThreadId>,
+    marker: PhantomData<*const T>,
+}
+
+unsafe impl<T: ?Sized> Send for WeakRef<T> {}
+unsafe impl<T: ?Sized> Sync for WeakRef<T> {}
+
+/// RAII for GWeakRef.
+#[derive(Debug)]
+struct WeakRefInner(gobject_ffi::GWeakRef);
+
+impl Drop for WeakRefInner {
     fn drop(&mut self) {
         unsafe {
-            gobject_ffi::g_weak_ref_clear(mut_override(&*self.0));
+            gobject_ffi::g_weak_ref_clear(mut_override(&self.0));
         }
     }
 }
 
-impl<T: IsA<Object> + ?Sized> Clone for WeakRef<T> {
-    fn clone(&self) -> Self {
-        unsafe {
-            let c = WeakRef(Box::new(mem::uninitialized()), PhantomData);
+/// An error returned by [`WeakRef::try_upgrade`](struct.WeakRef.html#method.try_upgrade).
+#[derive(Debug)]
+pub struct UpgradeError {
```

Comment:
We usually just use BoolError for such things here

Reply:
I was thinking about adding additional debug information like thread IDs, that
is why I used separate error type and left a room to change implementation
details without breaking API. Though, neither GLib thread IDs, nor Rust IDs
correspond to native threads IDs you could easily see in debugger, which I
think would be most useful. Thus for now I left error message without
additional details.